### PR TITLE
JAX: Add implementation & prepare release

### PR DESCRIPTION
**Pull request by AnirudhDagar**
*Description of changes:*

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

### CODE REVIEW

**Comment by astonzhang** [comment]
Quick one: can you double check why there're changes to d2l/torch, mxnet, tensorflow?

**Comment by AnirudhDagar** [comment]
>  Quick one: can you double check why there're changes to d2l/torch, mxnet, tensorflow?

That is expected, because I removed `LSTMScratch` save annotation in JAX branch for all the frameworks because it is not used in any other section.

Also the second change is `show_images` function, it has been just reordered because I changed that in [this commit](https://github.com/d2l-ai/d2l-en/pull/2381/commits/17641d17408903a9d98a29691ba52d3a5f99486a#diff-d43410e6811d33dd6f3f182680ae95b1d08b471f8f28848f55d5f69bba385e42L404-L434) while setting up the JAX D2L API.

Both these changes won't affect anything, just cleanups for all frameworks.

File: d2l/__init__.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 from d2l import mxnet as d2l  # Use MXNet as the backend
 from d2l import torch as d2l  # Use PyTorch as the backend
 from d2l import tensorflow as d2l  # Use TensorFlow as the backend
+from d2l import jax as d2l # Use Jax as the backend
```
Comment (astonzhang) [inline d2l/__init__.py:8]:
```suggestion
from d2l import jax as d2l  # Use Jax as the backend
```

File: chapter_attention-mechanisms-and-transformers/attention-pooling.md [hunk 2]
```diff
@@ -122,26 +138,32 @@ Recall attention pooling in :eqref:`eq_attention_pooling`. Let each validation f
 %%tab all
 def nadaraya_watson(x_train, y_train, x_val, kernel):
     dists = d2l.reshape(x_train, (-1, 1)) - d2l.reshape(x_val, (1, -1))
-    k = kernel(dists)  # Each column/row corresponds to each query/key
-    attention_w = k / k.sum(0)  # Normalization over keys for each query
+    k = d2l.astype(kernel(dists), d2l.float32)  # Each column/row corresponds to each query/key
```
Comment (astonzhang) [inline chapter_attention-mechanisms-and-transformers/attention-pooling.md:141]:
```suggestion
    # Each column/row corresponds to each query/key
    k = d2l.astype(kernel(dists), d2l.float32)
```

File: chapter_builders-guide/use-gpu.md [hunk 3]
```diff
@@ -434,17 +508,31 @@ with strategy.scope():
         tf.keras.layers.Dense(1)])
 ```
 
+```{.python .input}
+%%tab jax
+net = nn.Sequential([nn.Dense(1)])
+
+key1, key2 = jax.random.split(jax.random.PRNGKey(0))
+x = jax.random.normal(key1, (10,)) # Dummy input
```
Comment (astonzhang) [inline chapter_builders-guide/use-gpu.md:516]:
```suggestion
x = jax.random.normal(key1, (10,))  # Dummy input
```

File: chapter_builders-guide/use-gpu.md [hunk 4]
```diff
@@ -434,17 +508,31 @@ with strategy.scope():
...
+params = net.init(key2, x) # Initialization call
```
Comment (astonzhang) [inline chapter_builders-guide/use-gpu.md:517]:
```suggestion
params = net.init(key2, x)  # Initialization call
```

File: chapter_preliminaries/ndarray.md [hunk 5]
```diff
@@ -295,6 +317,14 @@ torch.randn(3, 4)
 tf.random.normal(shape=[3, 4])
 ```
 
+```{.python .input}
+%%tab jax
+# any call of a random function in JAX requires a key to be
```
Comment (astonzhang) [inline chapter_preliminaries/ndarray.md:322]:
```suggestion
# Any call of a random function in JAX requires a key to be 
```

File: chapter_preliminaries/ndarray.md [hunk 6]
```diff
@@ -369,6 +404,15 @@ X_var[1, 2].assign(9)
 X_var
 ```
 
+```{.python .input}
+%%tab jax
+# JAX arrays are immutable
+# `jax.numpy.ndarray.at` index update operators create
+# a new array with the corresponding modifications made
```
Comment (astonzhang) [inline chapter_preliminaries/ndarray.md:411]:
```suggestion
# JAX arrays are immutable. `jax.numpy.ndarray.at` index
# update operators create a new array with the corresponding
# modifications made
```

**Comment by astonzhang** [comment]
Merging since http://ci.d2l.ai/blue/organizations/jenkins/d2l-en/detail/PR-2381/2/pipeline passes (killed the other worker to save bandwidth for vol2 PRs)

**Pull request merged**